### DISCUSSION TITLE
Keycloak extension execute code on server startup

### DISCUSSION THREAD

**Discussion by cgeorgilakis** [keycloak/keycloak#24698] [Q&A]
We have an extension that we want to execute some tasks in server start up like Keycloak does in LegacyDatastoreProviderFactory.

In version 18 the following is working:
We use @ApplicationScoped annotation in class and we have the following method : 
` void onStart(@Observes StartupEvent ev) {
        logger.info("GroupManagement extension is starting...");
        //work only for quarkus
        QuarkusKeycloakSessionFactory instance = QuarkusKeycloakSessionFactory.getInstance();
        instance.init();
        try (KeycloakSession session = instance.create()) {
            TimerProvider timer = session.getProvider(TimerProvider.class); ...`

We want to migrate to version 22. In this version timer is null. So we can not execute startup tasks as previous.
What have been changed between version 18 and version 22? How can we execute  startup tasks in Keycloak extension?

**Comment by danielFesenmeyer** [comment]
It's probably better to listen to Keycloak's [PostMigrationEvent](https://github.com/keycloak/keycloak/blob/22.0.5/server-spi-private/src/main/java/org/keycloak/models/utils/PostMigrationEvent.java) instead, for example in a postInit method of an arbitrary ProviderFactory implementation.
This way, it should be guaranteed that Keycloak has been completely initialized, including the database (connection).

**Reply by cgeorgilakis** [reply]
I have seen it. However, I do not know how to listen it.
You could write : 
`  void onStart(@Observes  PostMigrationEvent ev) {`
?

**Reply by danielFesenmeyer** [reply]
You could use a RealmResourceProviderFactory with a dummy implementation (like always returning 404).
It should work like that:

```
public class DummyRealmResourceProviderFactory implements RealmResourceProviderFactory {

    @Override
    public RealmResourceProvider create(KeycloakSession keycloakSession) {
        return new DummyRealmResourceProvider(keycloakSession);
    }


    @Override
    public void postInit(KeycloakSessionFactory factory) {
        factory.register(event -> new InitListener(factory));
    }

    private static class InitListener implements ProviderEventListener {

        private final KeycloakSessionFactory sessionFactory;

        private InitListener(KeycloakSessionFactory sessionFactory) {
            this.sessionFactory = sessionFactory;
        }

        @Override
        public void onEvent(ProviderEvent providerEvent) {
            if (providerEvent instanceof PostMigrationEvent) {
                KeycloakSession session = sessionFactory.create();
                try {
                    // TODO: run your init code
                } finally {
                    session.close();
                }
                // unregister, cause further events are not needed
                sessionFactory.unregister(this);
            }
        }
    }
    
    ...
}
```
